src/data/bmg_plate.py

```diff
@@ -28,26 +31,72 @@ def __init__(self, barcode: str, plate_array: np.ndarray) -> None:
+        self.plate_array = plate_array.astype(np.float32)
+        self.pos = self.plate_array[:, -1]
+        self.neg = self.plate_array[:, -2]
+        (
+            self.std_pos,
+            self.std_neg,
+            self.mean_pos,
+            self.mean_neg,
+            self.z_factor,
+        ) = self.control_statistics(self.pos, self.neg)
+        self.pos_wo, self.outliers_pos = self.find_outliers(
+            self.pos, self.mean_pos, self.std_pos
         )
-
-    def control_statistics(self) -> None:
+        self.neg_wo, self.outliers_neg = self.find_outliers(
+            self.neg, self.mean_neg, self.std_neg
+        )
+        (
+            self.std_pos_wo,
+            self.std_neg_wo,
+            self.mean_pos_wo,
+            self.mean_neg_wo,
+            self.z_factor_wo,
+        ) = self.control_statistics(self.pos_wo, self.neg_wo)
+
+    def control_statistics(self, pos: np.ndarray, neg: np.ndarray) -> tuple:
         """
         Calculate statistic for control values e.g. two last columns of a plate
         """
-        pos = self.plate_array[:, -2]
-        neg = self.plate_array[:, -1]
-        self.std_pos = np.std(pos)
-        self.std_neg = np.std(neg)
-        self.mean_pos = np.mean(pos)
-        self.mean_neg = np.mean(neg)
-
-    def find_outliers(self):
-        # TODO
-        ...
+        std_pos = np.nanstd(pos)
+        std_neg = np.nanstd(neg)
+        mean_pos = np.nanmean(pos)
+        mean_neg = np.nanmean(neg)
+        z_factor = 1 - (3 * (std_pos + std_neg) / (mean_neg - mean_pos))
+        return std_pos, std_neg, mean_pos, mean_neg, z_factor
+
+    def find_outliers(
+        self, control: np.ndarray, control_mean: float, control_std: float
+    ) -> tuple:
+        """
+        Find outliers (max 2) in a given control array and assign them value of NaN.
+        The method is using standard deviation. In case of finding more than 2 outliers,
+        remove the most outling ones.
+
+        :param control: array containing control values (pos or neg)
+        :param control_mean: mean of given control array
+        :param control_std: std of given control array
+        :return: control array with outliers replaced to NaNs, tuple with outliers indices and values
+        """
+        cut_off = 3 * control_std
+        lower_limit = control_mean - cut_off
+        upper_limit = control_mean + cut_off
+        outliers = np.where((control > upper_limit) | (control < lower_limit))[0]
+        if len(outliers) == 0:
+            return control, [np.nan]
+        elif len(outliers) > 2:
+            while len(outliers) > 2:
+                max_o = np.max(outliers)
+                min_o = np.min(outliers)
+                if max_o - upper_limit > lower_limit - min_o:
+                    outliers = outliers[outliers != max_o]
+                else:
+                    outliers = outliers[outliers != min_o]
+        new_control = control.copy()
```

Comment:
Imo it'd be good for this function to just find the outliers and then based on these found outliers mutate the controls, so something like:

```py
controls = ...
outliers = find_outliers(controls)
controls[outliers] = np.nan
```

(furthermore the std and mean could be calculated within the function itself)

finally - is there a need to save the indices of outliers in the dataframe if we can find them in by searching for `nan`s in `plate_array`?

Reply:
The problem with indices is that we need to store original array in order to plot plate with outliers' values (`nan`s would change coloring). So my idea was to store indices of outliers in case they were needed and to allow for comparing outlier idx from df with colormap on plot.

Reply:
hmm and what if we, instead of putting these values in a separate column, to the `plate_data` added a binary matrix denoting if the value is an outlier? e.g.

```py

plate_array = [
   [0, 1, 2],
   [3, 99, -34],
   [5, 2, 1]
]

outlier_mask = [
   [0, 0, 0],
   [0, 1, 1],
   [0, 0, 0]
]

plate_data = np.array([plate_array, outlier_mask])
```

whilst the outlier matrix would be very sparse, I recon the memory footprint would be negligable (and this approach may prove more handy in vectorized numpy operations further on)

Reply:
differentiation between pos/neg outliers should be quite easy as negs will have much lesser value than pos

Reply:
with this approach if we further separate `plate_array` column to a 3D numpy array, we'd have two ready files to serialize as the original dataframe without `plate_array` and `outliers_pos/neg` will be parquetable

Reply:
Values of neg and pos depend on assay (neg may have greater values than pos),
I'm not sure why we would need an outlier mask, we just need to remove outliers and store the ids for a report (probably). There is no need to visualize them or handle in other way.
How about moving `plate_array`s to separate numpy array and leaving outliers ids in the original df? I could combine pos and neg outliers into one column.

Reply:
As we've discussed yesterday, both approaches have its merits, so its just boils down to how you'll be able to serialize the dataframe

I we are to stay with outliers in a separate column, you can convert the list into a string of ';' separated values - with that approach the df will be parquetable and we get the best of both worlds